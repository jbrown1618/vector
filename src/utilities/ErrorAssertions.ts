import { Matrix, MatrixData } from '../types/matrix/Matrix';
import { Vector } from '../types/vector/Vector';

/**
 * Throws an error of the 2d array `data` is not square - i.e. if
 * all of its rows do not have the same length
 */
export function assertRectangular(data: MatrixData<any>): void {
  if (data.length === 0) {
    return;
  }
  const rowSize = data[0].length;
  for (let i = 0; i < data.length; i++) {
    if (data[i].length !== rowSize) {
      throw Error('TODO - message');
    }
  }
}

/**
 * Throws an error if `matrix` is not square
 */
export function assertSquare(matrix: Matrix<any>): void {
  if (matrix.getNumberOfColumns() !== matrix.getNumberOfRows()) {
    throw new Error('TODO - message');
  }
}

/**
 * Throws an error if any member of `vectors` has a different length
 * from any other member.
 */
export function assertHomogeneous(vectors: Array<Vector<any>>): void {
  if (vectors.length === 0) {
    return;
  }
  const size = vectors[0].getDimension();
  for (let i = 0; i < vectors.length; i++) {
    if (vectors[i].getDimension() !== size) {
      throw Error('TODO - message');
    }
  }
}

/**
 * Throws an error if empty
 */
export function assertNonEmpty(values: any[]) {
  if (values.length > 0) {
    return;
  }
  throw Error('TODO - message');
}

/**
 * Throws an error if `index` is not a valid identifier for an entry in `vector`
 */
export function assertValidVectorIndex(vector: Vector<any>, index: number): void {
  assertValidIndex(index, vector.getDimension());
}

/**
 * Throws an error if `(rowIndex, colIndex)` is not a valid identifier for an entry in `matrix`
 */
export function assertValidMatrixIndex(
  matrix: Matrix<any>,
  rowIndex: number,
  colIndex: number
): void {
  assertValidIndex(rowIndex, matrix.getNumberOfRows());
  assertValidIndex(colIndex, matrix.getNumberOfColumns());
}

/**
 * Throws an error if `index` is not a valid index for an array-like object of length `size`
 */
export function assertValidIndex(index: number, size: number): void {
  if (index < 0 || index >= size) {
    throw Error('TODO - message');
  }
}
