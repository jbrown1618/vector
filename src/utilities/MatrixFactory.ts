import { Matrix } from '../Matrix';
import { VectorFactory } from './VectorFactory';
import { Vector } from '../Vector';

export class MatrixFactory {
  static identity(size: number): Matrix {
    if (size < 0) {
      throw Error();
    }

    const columns: Array<Vector> = [];
    for (let i = 0; i < size; i++) {
      columns.push(VectorFactory.elementaryVector(size, i));
    }
    return Matrix.fromColumnVectors(columns);
  }
}
