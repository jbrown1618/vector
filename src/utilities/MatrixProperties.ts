import { Matrix } from '..';

export function isSquare(matrix: Matrix<any>): boolean {
  return matrix.getNumberOfColumns() === matrix.getNumberOfRows();
}
