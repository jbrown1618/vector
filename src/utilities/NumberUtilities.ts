export function approximatelyEqual(
  first: number,
  second: number,
  epsilon: number = 0.000001
): boolean {
  if (Number.isNaN(first) || Number.isNaN(second)) {
    return false; // NaN should never equal itself
  }

  if (!Number.isFinite(first) || !Number.isFinite(second)) {
    return false; // Infinities do not equal themselves
  }

  return Math.abs(first - second) < epsilon;
}

export function mod(num: number, modulus: number): number {
  // Operator % yields negative results for negative numbers.
  return ((num % modulus) + modulus) % modulus;
}

export function random(min: number = 0, max: number = 1): number {
  return min + Math.random() * (max - min);
}

export function randomNormal(
  mean: number = 0,
  standardDeviation: number = 1,
  numberOfSamples: number = 10
): number {
  let total = 0;
  for (let i = 0; i < numberOfSamples; i++) {
    total += random();
  }

  const halfNumberOfSamples = numberOfSamples / 2;
  return mean + (standardDeviation * (total - halfNumberOfSamples)) / halfNumberOfSamples;
}

// memoize these results
const binomialResults: Map<number, Map<number, number>> = new Map();
const factorialResults: Map<number, number> = new Map();
factorialResults.set(0, 1);

export function binomial(n: number, k: number): number {
  if (!binomialResults.has(n)) {
    binomialResults.set(n, new Map());
  }
  const mapForA = binomialResults.get(n) as Map<number, number>;
  if (!mapForA.has(k)) {
    if (k > n) {
      mapForA.set(k, 0);
    } else {
      const numerator = factorial(n);
      const denominator = factorial(n - k) * factorial(k);
      const result = numerator / denominator;
      mapForA.set(k, result);
    }
  }

  return mapForA.get(k) as number;
}

export function factorial(n: number): number {
  if (!factorialResults.has(n)) {
    const result = n * factorial(n - 1);
    factorialResults.set(n, result);
  }

  return factorialResults.get(n) as number;
}
