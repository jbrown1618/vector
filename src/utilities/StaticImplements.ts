/**
 * Ensures at compile time that the static part of a class implements a certain interface.
 * Usage:
 *
 * ```
 * interface HasFoo {
 *   foo(): Bar;
 * }
 *
 * @StaticImplements<HasFoo>()
 * class HasStaticFoo {
 *   static foo(): Bar;
 * }
 * ```
 */
export function StaticImplements<T>() {
  // @ts-ignore
  return (constructor: T) => {};
}
