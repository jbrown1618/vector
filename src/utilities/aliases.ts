import { Vector } from '@lib/types/vector/Vector';
import { Matrix } from '@lib/types/matrix/Matrix';
import { NumberVector } from '@lib/types/vector/NumberVector';
import { NumberMatrix } from '@lib/types/matrix/NumberMatrix';

const vb = NumberVector.builder();
const mb = NumberMatrix.builder();

/**
 * Creates a new {@link Vector} of numbers.  See {@link VectorBuilder.fromArray}
 * @public
 */
export function vec(data: number[]): Vector<number> {
  return vb.fromArray(data);
}

/**
 * Creates a new {@link Matrix} of numbers.  See {@link MatrixBuilder.fromArray}
 * @public
 */
export function mat(data: number[][]): Matrix<number> {
  return mb.fromArray(data);
}

/**
 * Creates a new identity matrix of size `size`.  See {@link MatrixBuilder.identity}
 * @public
 */
export function eye(size: number): Matrix<number> {
  return mb.identity(size);
}

/**
 * Creates a new vector of all 1s.  See {@link VectorBuilder.ones}
 * @public
 */
export function ones(entries: number): Vector<number>;
/**
 * Creates a new matrix of all 1s.  See {@link MatrixBuilder.ones}
 * @public
 */
export function ones(rows: number, columns: number): Matrix<number>;
export function ones(rows: number, columns?: number): Matrix<number> | Vector<number> {
  return columns === undefined ? vb.ones(rows) : mb.ones(rows, columns);
}

/**
 * Creates a new vector of all 0s.  See {@link VectorBuilder.zeros}
 * @public
 */
export function zeros(entries: number): Vector<number>;
/**
 * Creates a new matrix of all 0s.  See {@link MatrixBuilder.zeros}
 * @public
 */
export function zeros(rows: number, columns: number): Matrix<number>;
export function zeros(rows: number, columns?: number): Matrix<number> | Vector<number> {
  return columns === undefined ? vb.zeros(rows) : mb.zeros(rows, columns);
}

/**
 * Creates a new matrix with the specified entries on the diagonal.  See {@link MatrixBuilder.diagonal}
 * @public
 */
export function diag(elements: number[]): Matrix<number> {
  return mb.diagonal(vec(elements));
}
