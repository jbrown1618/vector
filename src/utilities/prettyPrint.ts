import { Matrix } from '../types/matrix/Matrix';
import { Vector } from '../types/vector/Vector';

/**
 * @public
 */
export function prettyPrint(num: number): string;
export function prettyPrint<S>(vector: Vector<S>): string;
export function prettyPrint<S>(matrix: Matrix<S>): string;
export function prettyPrint<S>(input: number | Vector<S> | Matrix<S>): string {
  if (typeof input === 'number') {
    return prettyPrintNumber(input);
  }
  const isVector = !!(input as any).innerProduct;
  if (isVector) {
    return prettyPrintVector(input as Vector<S>);
  } else {
    return prettyPrintMatrix(input as Matrix<S>);
  }
}

function prettyPrintNumber(num: number): string {
  const str = num.toString();
  const fixed = num.toFixed(6);
  return str.length > fixed.length ? fixed : str;
}

function prettyPrintVector<S>(vector: Vector<S>): string {
  const ops = vector.ops();
  const width = getPrintWidth(vector);
  const prettyData = vector.toArray().map(val => ops.prettyPrint(val));
  return prettyData.map(value => '[ ' + spaces(width - value.length) + value + ' ]').join('\n');
}

function prettyPrintMatrix<S>(matrix: Matrix<S>): string {
  const ops = matrix.ops();
  const widthByCol = matrix.getColumnVectors().map(col => getPrintWidth(col));
  const prettyData = matrix.toArray().map(rowArr => rowArr.map(val => ops.prettyPrint(val)));

  return prettyData
    .map(
      rowArr =>
        '[ ' +
        rowArr.map((val, colIndex) => spaces(widthByCol[colIndex] - val.length) + val).join('  ') +
        ' ]'
    )
    .join('\n');
}

function getPrintWidth<S>(v: Vector<S>): number {
  return v
    .toArray()
    .map(val => v.ops().prettyPrint(val))
    .map(str => str.length)
    .reduce((max, curr) => (curr > max ? curr : max), 0);
}

function spaces(n: number): string {
  let sp = '';
  for (let i = 0; i < n; i++) {
    sp = sp + ' ';
  }
  return sp;
}
