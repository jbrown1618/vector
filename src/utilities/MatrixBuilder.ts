import { Matrix, MatrixConstructor, MatrixData, Vector } from '..';
import { assertHomogeneous, assertValidMatrixIndex } from './ErrorAssertions';
import { ScalarOperations } from '../types/scalar/ScalarOperations';

export type MatrixIndexFunction<ScalarType> = (i: number, j: number) => ScalarType;
export type MatrixEntryFunction<ScalarType> = (
  entry: ScalarType,
  i: number,
  j: number
) => ScalarType;

export class MatrixBuilder<
  ScalarType,
  VectorType extends Vector<ScalarType>,
  MatrixType extends Matrix<ScalarType>
> {
  private readonly _matrixConstructor: MatrixConstructor<ScalarType, VectorType, MatrixType>;

  constructor(matrixConstructor: MatrixConstructor<ScalarType, VectorType, MatrixType>) {
    this._matrixConstructor = matrixConstructor;
  }

  private ops(): ScalarOperations<ScalarType> {
    return this._matrixConstructor.ops();
  }

  /**
   * Returns a 0x0 matrix
   *
   * ```
   * matrixBuilder.empty(); // []
   * ```
   */
  empty(): MatrixType {
    return new this._matrixConstructor([]);
  }

  fromData(data: MatrixData<ScalarType>): MatrixType {
    return new this._matrixConstructor(data);
  }

  fromIndexFunction(
    numRows: number,
    numColumns: number,
    indexFunction: MatrixIndexFunction<ScalarType>
  ) {
    const data: MatrixData<ScalarType> = [];
    for (let i = 0; i < numRows; i++) {
      data[i] = [];
      for (let j = 0; j < numColumns; j++) {
        data[i][j] = indexFunction(i, j);
      }
    }
    return new this._matrixConstructor(data);
  }

  map(matrix: Matrix<ScalarType>, entryFunction: MatrixEntryFunction<ScalarType>) {
    return this.fromIndexFunction(matrix.getNumberOfRows(), matrix.getNumberOfColumns(), (i, j) =>
      entryFunction(matrix.getEntry(i, j), i, j)
    );
  }

  fromColumnVectors(columns: Vector<ScalarType>[]): MatrixType {
    assertHomogeneous(columns);
    const numberOfColumns = columns.length;
    if (numberOfColumns === 0) {
      return this.empty();
    }
    const numberOfRows = columns[0].getDimension();
    if (numberOfRows === 0) {
      return this.empty();
    }

    return this.fromIndexFunction(numberOfRows, numberOfColumns, (i, j) => columns[j].getEntry(i));
  }

  fromRowVectors(rows: Vector<ScalarType>[]): MatrixType {
    assertHomogeneous(rows);
    const numberOfRows = rows.length;
    if (numberOfRows === 0) {
      return this.empty();
    }
    const numberOfColumns = rows[0].getDimension();
    if (numberOfColumns === 0) {
      return this.empty();
    }

    return this.fromIndexFunction(numberOfRows, numberOfColumns, (i, j) => rows[i].getEntry(j));
  }

  /**
   * Returns a matrix of the specified dimensions, consisting of all zeros
   *
   * ```
   * matrixBuilder.zeros(2, 3);
   *
   * // [ 0 0 0 ]
   * // [ 0 0 0 ]
   * ```
   * @param {number} numberOfRows
   * @param {number} numberOfColumns
   * @returns {MatrixType}
   */
  zeros(numberOfRows: number, numberOfColumns: number = numberOfRows): MatrixType {
    return this.fromIndexFunction(numberOfRows, numberOfColumns, () => this.ops().zero());
  }

  /**
   * Returns a matrix of the specified dimensions, consisting of all ones
   *
   * ```
   * matrixBuilder.ones(2, 3);
   *
   * // [ 1 1 1 ]
   * // [ 1 1 1 ]
   * ```
   * @param {number} numberOfRows
   * @param {number} numberOfColumns
   * @returns {MatrixType}
   */
  ones(numberOfRows: number, numberOfColumns: number = numberOfRows): MatrixType {
    return this.fromIndexFunction(numberOfRows, numberOfColumns, () => this.ops().one());
  }

  /**
   * Returns a `size` x `size` identity matrix
   *
   * ```
   * matrixBuilder.identity(3);
   *
   * // [ 1 0 0 ]
   * // [ 0 1 0 ]
   * // [ 0 0 1 ]
   * ```
   * @param {number} size
   * @returns {MatrixType}
   */
  identity(size: number): MatrixType {
    return this.fromIndexFunction(
      size,
      size,
      (i, j) => (i === j ? this.ops().one() : this.ops().zero())
    );
  }

  /**
   * Returns a square diagonal matrix whose diagonal entries come from `diagonalEntries`
   *
   * ```
   * const diagonalEntries = NumberVector.fromValues(1, 2, 3);
   * matrixBuilder.diagonal(diagonalEntries);
   *
   * // [ 1 0 0 ]
   * // [ 0 2 0 ]
   * // [ 0 0 3 ]
   * ```
   * @param {VectorType} diagonalEntries
   * @returns {MatrixType}
   */
  diagonal(diagonalEntries: VectorType): MatrixType {
    const size = diagonalEntries.getDimension();
    return this.fromIndexFunction(
      size,
      size,
      (i, j) => (i === j ? diagonalEntries.getEntry(i) : this.ops().zero())
    );
  }

  /**
   * Returns a square tridiagonal matrix whose diagonal entries correspond to the entries of
   * `diagonalEntries`, whose entries in the left-off-diagonal correspond to the entries
   * of `leftEntries`, and whose entries in the right-off-diagonal correspond fo the
   * entries of `rightEntries`.
   * The off-diagonals must have one fewer entry than the diagonal.
   * Throws an error if the dimensions are not correct.
   *
   * ```
   * const leftEntries = NumberVector.fromEntries(1, 2);
   * const diagonalEntries = NumberVector.fromEntries(3, 4, 5);
   * const rightEntries = NumberVector.fromEntries(6, 7);
   *
   * matrixBuilder.tridiagonal(leftEntries, diagonalEntries, rightEntries);
   *
   * // [ 3 6 0 ]
   * // [ 1 4 7 ]
   * // [ 0 2 5 ]
   * ```
   *
   * @param {VectorType} leftEntries
   * @param {VectorType} diagonalEntries
   * @param {VectorType} rightEntries
   * @returns {MatrixType}
   */
  tridiagonal(
    leftEntries: Vector<ScalarType>,
    diagonalEntries: Vector<ScalarType>,
    rightEntries: Vector<ScalarType>
  ): MatrixType {
    const size = diagonalEntries.getDimension();
    const hasSizeMismatch =
      leftEntries.getDimension() !== size - 1 || rightEntries.getDimension() !== size - 1;

    if (hasSizeMismatch) {
      throw Error('');
    }

    return this.fromIndexFunction(size, size, (i, j) => {
      if (i === j) {
        return diagonalEntries.getEntry(i);
      } else if (i === j + 1) {
        return leftEntries.getEntry(j);
      } else if (i === j - 1) {
        return rightEntries.getEntry(i);
      } else {
        return this.ops().zero();
      }
    });
  }

  /**
   * Returns a new matrix consisting of `left` and `right` next to one another.
   * Throws an error of `left` and `right` do not have the same number of rows.
   *
   * ```
   * const left = matrixBuilder.ones(2, 3);
   * const right = matrixBuilder.zeros(2, 4);
   *
   * matrixBuilder.augment(left, right);
   *
   * // [ 1 1 1 0 0 0 0 ]
   * // [ 1 1 1 0 0 0 0 ]
   * ```
   * @param {MatrixType} left
   * @param {MatrixType} right
   * @returns {MatrixType}
   */
  augment(left: Matrix<ScalarType>, right: Matrix<ScalarType>): MatrixType {
    if (left.getNumberOfRows() !== right.getNumberOfRows()) {
      throw Error('Dimension mismatch!');
    }

    return this.fromColumnVectors([...left.getColumnVectors(), ...right.getColumnVectors()]);
  }

  /**
   * Returns a new matrix consisting of `top` and `bottom` on top of one another.
   * Throws an error if `top` and `bottom` do not have the same number of columns.
   *
   * ```
   * const top = matrixBuilder.ones(2, 3);
   * const bottom = matrixBuilder.zeros(1,3);
   *
   * matrixBuilder.stack(top, bottom);
   *
   * // [ 1 1 1 ]
   * // [ 1 1 1 ]
   * // [ 0 0 0 ]
   * ```
   * @param {MatrixType} top
   * @param {MatrixType} bottom
   * @returns {MatrixType}
   */
  private stack(top: Matrix<ScalarType>, bottom: Matrix<ScalarType>): MatrixType {
    if (top.getNumberOfColumns() !== bottom.getNumberOfColumns()) {
      throw Error('Dimension mismatch!');
    }

    return this.fromRowVectors([...top.getRowVectors(), ...bottom.getRowVectors()]);
  }

  /**
   * Returns a single matrix consisting of a grid of matrices combined together.
   * Throws an error if any of the dimensions are incompatible.
   *
   * ```
   * const upperLeft = matrixBuilder.ones(1, 1);
   * const upperRight = matrixBuilder.zeros(1, 2);
   * const lowerLeft = matrixBuilder.zeros(2, 1);
   * const lowerRight = matrixBuilder.ones(2, 2);
   *
   * const grid = [
   *   [ upperLeft, upperRight ],
   *   [ lowerLeft, lowerRight ]
   * ];
   *
   * matrixBuilder.flatten(grid);
   *
   * // [ 1 0 0 ]
   * // [ 0 1 1 ]
   * // [ 0 1 1 ]
   * ```
   * @param {Array<Array<MatrixType>>} grid
   * @returns {MatrixType}
   */
  flatten(grid: Array<Array<MatrixType>>): MatrixType {
    if (grid.length === 0 || grid[0].length === 0) {
      return this.empty();
    }

    return grid
      .map(gridRow => {
        return gridRow.reduce((accumulator: MatrixType, gridEntry: MatrixType) => {
          return this.augment(accumulator, gridEntry);
        });
      })
      .reduce((accumulator: MatrixType, row: MatrixType) => {
        return this.stack(accumulator, row);
      });
  }

  /**
   * Returns a new matrix consisted of repetitions of a smaller matrix.
   *
   * ```
   * const I = matrixBuilder.identity(2);
   * matrixBuilder.repeat(I, 1, 2);
   *
   * // [ 1 0 1 0 ]
   * // [ 0 1 0 1 ]
   * ```
   * @param {MatrixType} matrix
   * @param {number} rows
   * @param {number} columns
   * @returns {MatrixType}
   */
  repeat(matrix: MatrixType, rows: number, columns: number) {
    const grid: Array<Array<MatrixType>> = [];

    for (let i = 0; i < rows; i++) {
      grid[i] = [];
      for (let j = 0; j < columns; j++) {
        grid[i][j] = matrix;
      }
    }

    return this.flatten(grid);
  }

  /**
   * Returns a new matrix based on a rectangular slice of a larger matrix
   *
   * ```
   * const matrix = matrixBuilder.identity(4);
   * matrixBuilder.slice(matrix, 2, 2, 3, 4);
   *
   * // [  1  0  0  0  ]
   * // [  0  1* 0* 0* ]  =>  [ 1 0 0 ]
   * // [  0  0* 1* 0* ]      [ 0 1 0 ]
   * // [  0  0  0  1  ]
   * ```
   *
   * @param {MatrixType} matrix  The original matrix
   * @param {number} rowStartIndex  The (inclusive) first row of the slice
   * @param {number} columnStartIndex  The (inclusive) first column of the slice
   * @param {number} rowEndIndex  The (exclusive) last row of the slice
   * @param {number} columnEndIndex  The (exclusive) last column of the slice
   * @returns {MatrixType}
   */
  slice(
    matrix: Matrix<ScalarType>,
    rowStartIndex: number = 0,
    columnStartIndex: number = 0,
    rowEndIndex: number = matrix.getNumberOfRows(),
    columnEndIndex: number = matrix.getNumberOfColumns()
  ): MatrixType {
    if (rowStartIndex > rowEndIndex || columnStartIndex > columnEndIndex) {
      throw Error('start index must be less than end index');
    }

    if (rowStartIndex < 0 || columnStartIndex < 0) {
      throw Error('indices must be positive');
    }

    if (rowEndIndex > matrix.getNumberOfRows() || columnEndIndex > matrix.getNumberOfColumns()) {
      throw Error('index out of bounds');
    }

    const data: MatrixData<ScalarType> = [];
    let newRowIndex = 0;
    for (let i = rowStartIndex; i < rowEndIndex; i++) {
      data[newRowIndex] = [];
      let newColumnIndex = 0;
      for (let j = columnStartIndex; j < columnEndIndex; j++) {
        data[newRowIndex][newColumnIndex] = matrix.getEntry(i, j);
        newColumnIndex++;
      }
      newRowIndex++;
    }

    return this.fromData(data);
  }

  /**
   * Returns a new matrix with all entries in row `rowToExclude` and in
   * column `columnToExclude` removed.
   *
   * ```
   * const I = matrixBuilder.identity(4);
   * matrixBuilder.slice(I, 1, 2)
   *
   * // [  1  0  0* 0  ]
   * // [  0* 1* 0* 0* ]  =>  [ 1 0 0 ]
   * // [  0  0  1* 0  ]      [ 0 0 0 ]
   * // [  0  0  0* 1  ]      [ 0 0 1 ]
   * ```
   * @param {MatrixType} matrix
   * @param {number} rowToExclude
   * @param {number} columnToExclude
   * @returns {MatrixType}
   */
  exclude(matrix: MatrixType, rowToExclude: number, columnToExclude: number): MatrixType {
    assertValidMatrixIndex(matrix, rowToExclude, columnToExclude);

    const data: MatrixData<ScalarType> = [];
    for (let i = 0; i < matrix.getNumberOfRows(); i++) {
      if (i < rowToExclude) {
        data[i] = [];
        for (let j = 0; j < matrix.getNumberOfColumns(); j++) {
          if (j < columnToExclude) {
            data[i][j] = matrix.getEntry(i, j);
          } else if (j > columnToExclude) {
            data[i][j - 1] = matrix.getEntry(i, j);
          }
        }
      } else if (i > rowToExclude) {
        data[i - 1] = [];
        for (let j = 0; j < matrix.getNumberOfColumns(); j++) {
          if (j < columnToExclude) {
            data[i - 1][j] = matrix.getEntry(i, j);
          } else if (j > columnToExclude) {
            data[i - 1][j - 1] = matrix.getEntry(i, j);
          }
        }
      }
    }

    return this.fromData(data);
  }
}
