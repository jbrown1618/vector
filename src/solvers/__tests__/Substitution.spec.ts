import { mat, vec } from '../../utilities/aliases';
import {
  backwardSubstituteAugmentedMatrix,
  forwardSubstituteAugmentedMatrix,
  solveByBackwardSubstitution,
  solveByForwardSubstitution,
} from '../Substitution';

describe('Substitution', () => {
  describe('solveByForwardSubstitution', () => {
    test('solves a lower-triangular system with a unique solution', () => {
      const L = mat([
        [2, 0, 0],
        [2, 1, 0],
        [3, 3, 2],
      ]);
      const b = vec([3, 2, 3]);

      const x = solveByForwardSubstitution(L, b)!;

      expect(x).toStrictEqual(vec([1.5, -1, 3 / 4]));
      expect(L.apply(x)).toStrictEqual(b);
    });

    test('returns undefined for underdetermined system', () => {
      expect(
        solveByForwardSubstitution(
          mat([
            [0, 0, 0],
            [0, 0, 0],
            [1, 1, 1],
          ]),
          vec([0, 0, 3])
        )
      ).toBeUndefined();

      expect(
        solveByForwardSubstitution(
          mat([
            [0, 0, 0],
            [1, 0, 0],
            [0, 2, 0],
          ]),
          vec([0, 0, 0])
        )
      ).toBeUndefined();
    });

    test('returns undefined for an overdetermined system', () => {
      expect(
        solveByForwardSubstitution(
          mat([
            [0, 0],
            [1, 1],
          ]),
          vec([1, 1])
        )
      ).toBeUndefined();
    });

    test('throws for non-square matrices', () => {
      expect(() => {
        solveByForwardSubstitution(
          mat([
            [1, 0],
            [0, 1],
            [0, 1],
          ]),
          vec([1, 2, 2])
        );
      }).toThrow();

      expect(() => {
        solveByForwardSubstitution(
          mat([
            [1, 0, 0],
            [0, 1, 0],
          ]),
          vec([1, 2])
        );
      }).toThrow();
    });

    test('handles the trivial case', () => {
      expect(solveByForwardSubstitution(mat([]), vec([]))).toStrictEqual(vec([]));
    });
  });

  describe('forwardSubstituteAugmentedMatrix', () => {
    test('solves a lower-triangular system with a unique solution', () => {
      const L = mat([
        [2, 0, 0, 3],
        [2, 1, 0, 2],
        [3, 3, 2, 3],
      ]);
      const x = forwardSubstituteAugmentedMatrix(L);

      expect(x).toStrictEqual(vec([1.5, -1, 3 / 4]));
    });

    test('returns undefined for an underdetermined lower-triangular system', () => {
      const L = mat([
        [0, 0, 0, 0],
        [0, 0, 0, 0],
        [1, 1, 1, 3],
      ]);
      const x = forwardSubstituteAugmentedMatrix(L);

      expect(x).toBeUndefined();
    });

    test('returns undefined for an overdetermined lower-triangular system', () => {
      const L = mat([
        [0, 0, 1],
        [1, 1, 1],
      ]);
      const x = forwardSubstituteAugmentedMatrix(L);

      expect(x).toBeUndefined();
    });
  });

  describe('solveByBackwardSubstitution', () => {
    test('solves an upper-triangular system with a unique solution', () => {
      const U = mat([
        [2, 2, 3],
        [0, 1, 2],
        [0, 0, 2],
      ]);
      const b = vec([3, 2, 3]);
      const x = solveByBackwardSubstitution(U, b)!;

      expect(x).toStrictEqual(vec([0.25, -1, 1.5]));
      expect(U.apply(x)).toStrictEqual(b);
    });

    test('returns undefined for an underdetermined upper-triangular system', () => {
      let U = mat([
        [1, 1, 1],
        [0, 0, 0],
        [0, 0, 0],
      ]);
      let b = vec([3, 0, 0]);
      let x = solveByBackwardSubstitution(U, b);

      expect(x).toBeUndefined();

      U = mat([
        [0, 1, 0],
        [0, 0, 2],
        [0, 0, 0],
      ]);
      b = vec([0, 0, 0]);
      x = solveByBackwardSubstitution(U, b);

      expect(x).toBeUndefined();
    });

    test('returns undefined for an overdetermined upper-triangular system', () => {
      const U = mat([
        [1, 1],
        [0, 0],
      ]);
      const b = vec([1, 1]);
      const x = solveByBackwardSubstitution(U, b);

      expect(x).toBeUndefined();
    });

    test('throws for non-square matrices', () => {
      expect(() =>
        solveByBackwardSubstitution(
          mat([
            [1, 0],
            [0, 1],
            [0, 0],
          ]),
          vec([1, 2, 0])
        )
      ).toThrow();

      expect(() =>
        solveByBackwardSubstitution(
          mat([
            [1, 0, 0],
            [0, 1, 0],
          ]),
          vec([1, 2])
        )
      ).toThrow();
    });

    test('handles the trivial case', () => {
      expect(solveByBackwardSubstitution(mat([]), vec([]))).toStrictEqual(vec([]));
    });
  });

  describe('backwardSubstituteAugmentedMatrix', () => {
    test('solves an upper-triangular system with a unique solution', () => {
      const U = mat([
        [2, 2, 3, 3],
        [0, 1, 2, 2],
        [0, 0, 2, 3],
      ]);
      const x = backwardSubstituteAugmentedMatrix(U);

      expect(x).toStrictEqual(vec([0.25, -1, 1.5]));
    });

    test('returns undefined for an underdetermined upper-triangular system', () => {
      const U = mat([
        [1, 1, 1, 3],
        [0, 0, 0, 0],
        [0, 0, 0, 0],
      ]);
      const x = backwardSubstituteAugmentedMatrix(U);
      expect(x).toBeUndefined();
    });

    test('returns undefined an overdetermined upper-triangular system', () => {
      const U = mat([
        [1, 1, 1],
        [0, 0, 1],
      ]);
      const x = backwardSubstituteAugmentedMatrix(U);
      expect(x).toBeUndefined();
    });
  });
});
