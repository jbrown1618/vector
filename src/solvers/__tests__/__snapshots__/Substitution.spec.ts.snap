// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Substitution solveByBackwardSubstitution solves a random 50x50 upper-triangular system 1`] = `
FloatVector {
  "_data": Float64Array [
    -75.4055043924601,
    -211.76477054890202,
    106.78005911977984,
    101.57645681560861,
    -193.58533420942268,
    18.068856858020883,
    10.533730242555082,
    -235.79665456068133,
    -6.0242855853844635,
    -96.75981018341813,
    105.17609804469038,
    33.69131171319699,
    77.40310018415697,
    -231.350506886773,
    118.51710515031839,
    -244.70848421711324,
    -853.5444570133816,
    -57.88644530902292,
    -104.18783355724102,
    -212.9422277001505,
    -82.29809205396089,
    342.44756857888865,
    12.131151151439417,
    -63.60305145001421,
    -27.939123406335312,
    544.5024309846526,
    -58.13989143260359,
    -129.64939818106794,
    23.812231734396317,
    -169.8398499799322,
    -148.17835246593955,
    114.85082451843427,
    -82.16483769481273,
    98.99413615411319,
    65.0630042313794,
    -248.44185001631953,
    -414.2497239280466,
    27.617398900783446,
    -134.19621171387428,
    225.25057760825652,
    121.2936257529631,
    -489.0659422537312,
    108.26959888802612,
    85.5212708134847,
    58.34539160769554,
    83.07030219703749,
    -6.581168210878951,
    468.38994294137586,
    20.70464601854587,
    121.24179831536522,
  ],
}
`;

exports[`Substitution solveByForwardSubstitution solves a random 50x50 lower-triangular matrix 1`] = `
FloatVector {
  "_data": Float64Array [
    -77.14397065945583,
    -38.00080611520708,
    2759.842151625131,
    -2969.3911483311717,
    -233.52005691369718,
    -720.462181056047,
    -6594.122326160257,
    7127.486036634982,
    3442.809203877073,
    -4274.110466691362,
    5921.658396998813,
    197919.3143701788,
    -4150110.731320688,
    4189805.8561148015,
    -776114.4854102368,
    2424877.077914972,
    -2320823.9508698313,
    28395252.11112745,
    3461500.136695737,
    -78882022.6181815,
    -31305854.51209573,
    87111418.33625871,
    362093614.99351656,
    -63064457.182486616,
    -102439832.6658007,
    -382400188.70842016,
    338359869.1664802,
    -457277709.6147028,
    5155009582.956564,
    -329949798.726457,
    -6355603477.434715,
    3212673990.066189,
    6077019362.880722,
    -12672353291.880222,
    -2551516148.3273487,
    7283724750.076708,
    11275597450.25983,
    -22658542687.922623,
    -21596654513.55598,
    4535436000.308833,
    -26687541939.65937,
    28529450749.839626,
    3629169014.8053436,
    -3695292538.1543503,
    3021107380.1066437,
    86994458170.49428,
    -98305631512.38417,
    -37347325861.91859,
    -288633179245.37213,
    291582231293.62396,
  ],
}
`;
